src/charm.py

```diff
@@ -19,6 +24,183 @@ class KratosCharm(CharmBase):
+        self.framework.observe(self.database.on.database_created, self._on_database_changed)
+        self.framework.observe(self.database.on.endpoints_changed, self._on_database_changed)
+
+    @property
+    def _pebble_layer(self) -> Layer:
+        pebble_layer = {
+            "summary": "kratos layer",
+            "description": "pebble config layer for kratos",
+            "services": {
+                self._container_name: {
+                    "override": "replace",
+                    "summary": "Kratos Operator layer",
+                    "startup": "enabled",
+                    "command": "kratos serve all --config /etc/config/kratos.yaml",
+                }
+            },
+            "checks": {
+                "ready": {
+                    "override": "replace",
+                    "http": {"url": "http://localhost:4434/admin/health/ready"},
+                },
+                "alive": {
+                    "override": "replace",
+                    "http": {"url": "http://localhost:4434/admin/health/alive"},
+                },
+            },
+        }
+        return Layer(pebble_layer)
+
+    @property
+    def _config(self) -> str:
+        db_info = self._get_database_relation_info()
+        config = {
+            "log": {"level": "trace"},
+            "identity": {
+                "default_schema_id": "default",
+                "schemas": [{"id": "default", "url": f"file://{self._identity_schema_file_path}"}],
+            },
+            "selfservice": {
+                "default_browser_return_url": "http://127.0.0.1:9999/",
+                "flows": {
+                    "registration": {
+                        "enabled": True,
+                        "ui_url": "http://127.0.0.1:9999/registration",
+                    }
+                },
+            },
+            "dsn": f"postgres://{db_info['username']}:{db_info['password']}@{db_info['endpoints']}/postgres",
+            "courier": {
+                "smtp": {
+                    # TODO: dynamic connection uri through charm config
+                    "connection_uri": "smtps://test:test@mailslurper:1025/?skip_ssl_verify=true"
+                }
+            },
+        }
+
+        return yaml.dump(config)
+
+    def _update_layer(self) -> None:
+        """Updates the Pebble configuration layer and kratos config if changed."""
+        # Get current layer
+        current_layer = self._container.get_plan()
+        # Create a new config layer
+        new_layer = self._pebble_layer
+
+        if current_layer.services != new_layer.services:
+            self.unit.status = MaintenanceStatus("Applying new pebble layer")
+            self._container.push(self._config_file_path, self._config, make_dirs=True)
+            self._container.add_layer(self._container_name, new_layer, combine=True)
+            logger.info("Pebble plan updated with new configuration, replanning")
+            self._container.replan()
```

Comment:
I have seen that in other charms this is written like this:

```python
        self._container.add_layer(self.container_name, new_layer, combine=True)
        self._container.autostart()
```

Is there a difference between the two? Do we need to check for diffs in the layers?

Reply:
I honestly don't know. I posted a question on the public mattermost.